### PR TITLE
Include sealed traits and classes as elements of Sum Mirrors

### PR DESCRIPTION
**Pull request by odersky**
Addresses feature request #161

https://github.com/lampepfl/dotty-feature-requests/issues/161

This is just the immediate fix. We need more tests and use cases to see whether it works. @dwijnand if we can get some corrobation by Wednesday we might be able to roll this into 3.0. Otherwise it will have to wait.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/SymUtils.scala [hunk 1]
```diff
@@ -95,7 +95,7 @@ object SymUtils:
     *   - none of its children are anonymous classes
     *   - all of its children are addressable through a path from the parent class
     *     and also the location of the generated mirror.
-    *   - all of its children are generic products or singletons
+    *   - all of its children are generic products, singletons, or sealed traits ar classes
```
Comment (julienrf) [inline compiler/src/dotty/tools/dotc/transform/SymUtils.scala:98]:
```suggestion
    *   - all of its children are generic products, singletons, or sealed traits or classes
```

File: tests/i9276.scala [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+case class C() extends A
+
+sealed trait B extends A
+
+case class D() extends B
+
+@main def Test =
+  summon[Mirror.Of[A]]
+  summon[Mirror.Of[B]]
```
Comment (julienrf) [inline tests/i9276.scala:13]: Can we maybe add another test like the following?

~~~ scala
val mirrorA = summon[Mirror.Of[A]]
summon[mirrorA.MirroredElemTypes <:< (C *: B *: EmptyTuple)]
~~~

**Review by julienrf — COMMENTED** [review]
Thanks for the quick fix!

**Comment by odersky** [comment]
As the failure of the Akka CB shows that I have overlooked something. The trait hierarchy can be an arbitrary graph. Consider:
```scala
sealed trait A
sealed trait B extends A
sealed trait C extends A
case class D() extends B, C
case class E() extends B, C
```
Now if we generate the ordinal method, we get unreachable cases. That's what we observe when trying to build akka.

**Comment by julienrf** [comment]
>     1. Continue with #11686. This means that mirrors will show elements that are not in the children of the base trait. This is risky since it might break some assumptions of code that does typeclass derivation.

Which assumptions do you think about?

**Comment by odersky** [comment]
> Which assumptions do you think about?

For instance, if I have typeclass derivation code that consults the children annotation at runtime I get a mismatch between that and what the mirror claims. Or if the code assumes that all mirror elements have the sum as direct parent type.

**Comment by lrytz** [comment]
(Dale is OOO this week)

**Comment by japgolly** [comment]
> if we can get some corrobation by Wednesday we might be able to roll this into 3.0
> (Dale is OOO this week)

I'm happy to put my hand up for that. Let me know what you need!

**Comment by japgolly** [comment]
Regarding the approach to implementation, I have another one to add to the 3 ideas above.

(4) Have two means of summoning a mirror, both with simple algorithms:

1) Gather all product children transitively. This is the most commonly useful case so could be the default for the existing `Mirror.Of`. In the Akka example above:

```scala
summon[Mirror.Of[A]].MirroredElemTypes =:= (C, D)
summon[Mirror.Of[B]].MirroredElemTypes =:= (D)
```

2) Gather sub-types directly and non-transitively, regardless of whether the sub-type is a product or a sum. We could give this a different type `Mirror.DirectlyOf` and have users explicitly opt-in to this algorithm. In the Akka example above:

```scala
summon[Mirror.DirectlyOf[A]].MirroredElemTypes =:= (B, C, D)
summon[Mirror.DirectlyOf[B]].MirroredElemTypes =:= (D)
```
In my experience with automatic derivation, it's always one of the above two I want. Which one depends on what I'm deriving.

**Comment by julienrf** [comment]
@japgolly Interesting. I’m not sure whether we can change this now but I agree with your analysis: those two ways of retrieving the subclasses are the most common in my experience as well.

**Comment by odersky** [comment]
I think it's probably best for now to continue with #11686. But since Dale is OOO this weak there's no chance for it to make it into RC2. In my opinion, even if we could work on this right now, it would be too big a change to make so short before the release date.

**Pull request closed without merging**